### DISCUSSION TITLE
Help Changing Sonic's Default Abilities/Making New Character

### DISCUSSION THREAD

**Discussion by shifturret** [Techokami/SonicWorldsNext#156] [Q&A]
is there any way that i can make Sonic's default air action change from the drop dash to the fire shield attack?, if not, can i make a new character with this change? thanks in advance.

**Comment by JimmyThrush** [comment]
You want to change Normal.gd, which you should find by default in Scripts/Player/States. In the _process function of that script, you'll find the relevant logic for all the characters' air abilities, including Sonic's instashield, drop dash, and shield abilities. Just replace the instashield/drop dash code with the fire shield code, making the relevant modifications to the animation.

**Reply by shifturret** [reply]
I'm still learning gdscript and code as a whole, what should i look for in the script?

**Reply by JimmyThrush** [reply]
Sorry, I got the wrong script; it's Air.gd, not Normal.gd. In that script, you'll see comments listing the characters' abilities: "insta shield", "fire shield action", "tails flight", "knuckles gliding", etc. From there you'll have to figure it out on your own, that's everything you'll need to replace Sonic's default action

**Reply by shifturret** [reply]
I found it and changed the instashield action to the fire shield action, thanks! another question though, how do i change sonic's sprites, and if i change the sprites to a larger one will that cause problems with the hitbox?

**Reply by DimensionWarped** [reply]
Judging from experience, going outside the norm on character sprite size is quite tricky and is going to require realigning a lot of collision masks, raycast emitters, and even likely manual code changes. It's tricky. And that's of course after you align you make sure your sprite sheet is the right size, has the right number of subdivisions, and make sure your animationplayer still has everything aligned correctly too.

It's a big lift. No question. Plus you know... any larger sprite you might find/make is probably going to have a dearth of animations to use anyway. Don't know if I'd recommend it.

Of course you can always just edit the sprite sheet in place, understanding that you'll be constrained to the current limit of pixels per frame if you do so.

**Reply by NeonSRB2** [reply]
To make a seperate character in the code, Sonic Worlds Next doesn't make it easy like adding a new state, as several pieces of code just sift through a table indexed by the character variable, but here's about everything.

First is to add your character to the list of characters in Player.gd and Global.gd, which look like this:
`enum CHARACTERS {NONE,SONIC,TAILS,KNUCKLES,AMY} # global.gd`
`enum CHARACTERS {SONIC, TAILS, KNUCKLES, AMY} # player.gd`
the order of these is important, so you'll want to just add your character to the end of the list, name him whoever you like in the code, just make sure it's consistent and makes enough sense where you can type it in without having to look at it again.
Also Global.gd is an autoload script that isn't referenced in any scene, you can find it in either project settings or in Scripts/Global, but know that you only really have to do this once to get it into the editor's script list.

Now while your character exists, player.gd itself has many more character checks strewn throughout, though most of them are'nt really that important in my opinion, so I'll go through them here.
```
match (character):
			CHARACTERS.SONIC:
				# shader texture sizes need to be to the power of 2
				playerPal.set_shader_parameter("amount",4)
				playerPal.set_shader_parameter("palRows",16)
				playerPal.set_shader_parameter("row",0)
				playerPal.set_shader_parameter("paletteTexture",load("res://Graphics/Palettes/SuperSonicPal.png"))
...
```
This is for the super form, which I've never messed too much with so take me with a grain of salt for anything that doesn't work past not crashing. Make an entry for your character like: `CHARACTERS.SURGE` or whatever you put as the name before. As for the texture, make sure the first row is the colors you want to replace, and then any subsequent rows be whatever colors you want those to switch to. The actual animation part is handled in the character scene later, but copy one of these entries, and amount is going to be the horizontal width of the texture, and palRows is the length and height. change the paletteTexture's path to your texture next and you're good for now, you can drag the file from the filesystem tab to the script window to get the path.
```
func determine_physics():
	# get physics from character
	match (character):
		CHARACTERS.SONIC:
			if isSuper:
				return 4 # Super Sonic
			elif shoeTime > 0:
				return 3 # Shoes
			return 0 # Sonic
...
```
As the function suggests, this determines the table of physics each character uses... despite each entry being litterally identical...
Just copy Sonic's again, and I mean it completley this time, unless you want a lower jump height like Knuckles.

Now thats the stuff Player.gd will contain, and no other script is like this thankfully, but that doesn't mean you don't have to worry about them at all. This next step is to add your character to a few more things to stop some crashes and make him selectable.

HUD.gd (String Name, provide an extra listing after amy for your character)
Normal.gd (Idle anim list, just copy sonic's until you need to modify it.)
CharacterSelect.gd (Make your character selectable)

Just add a new entry for the tables at the top of each script, but for CharacterSelect.gd you'll need to make a bit more code, like how the level select works. By adding your string name to the list, you make an option to switch to, but you need to go down to another bit to set the character, like here:
```
match(characterID):
				0: # Sonic and Tails
					Global.PlayerChar1 = Global.CHARACTERS.SONIC
					Global.PlayerChar2 = Global.CHARACTERS.TAILS
...
```
I think you get the deal by now, but take note of that number at the top, since that refers to the actual order of those string listings you add. You can add Knuckles and Tails easily like this too if you feel it. The Global.CHARACTERS.YOU is of course, your characters' version of that.

Now your character should load and only crash when you use an ability, which I'm not helping you with since I have no idea what this character needs there aside from that Air.gd entry for... something, and I gotta stop typing this sometime.

But I do know that you need sprites, so let's do that. In Player.gd you may notice a referenced Node2D simply called "Sonic", this is the sprites, super palette animation, and also regular animations too. These are stored in a wierd place, Graphics/Players/PlayerAnimations, and despite the fact that you're in the Graphics folder there are in fact the character sprites there. Copy Sonic's scene and rename it to whatever, like your character's name or something. Once you've done that, try not to touch it _too_ much, as it's probably going to be a pain for you to fix stuff here. If your character fits in the 48x48 pixel area each of sonic's frames have to reside in, you should just copy the sheet's .png and swap the texture field of the Sprite2d node with your sheet. If you need to mess with how many frames each animation has, just go look around in that AnimationPlayer node, shouldn't be too hard to figure it out from there. Oh, and since I'm talking about super palettes now, the rows property you see in the animations of that "SuperPalette node" is the row that your character uses for the respective palette animation, edit as you please.

So, the scene is here, now we gotta implement it into Player.gd to make it switch, but first you need to make a variable to house the new scene.

```
var tailsAnimations = preload("res://Graphics/Players/PlayerAnimations/Tails.tscn")
var knucklesAnimations = preload("res://Graphics/Players/PlayerAnimations/Knuckles.tscn")
var amyAnimations = preload("res://Graphics/Players/PlayerAnimations/Amy.tscn")
```
As when I had you swap the texture's path to the image you made, make a new entry that looks something like this:
`var surgeAnimations = preload("res://Graphics/Players/PlayerAnimations/Surge.tscn")`
Basically just as long as its named similarly for readibility and the path is correct, it should work.

Next, is the big boy.
```
CHARACTERS.TAILS:
			# Set sprites
			currentHitbox = HITBOXESTAILS
			get_node("Sonic").name = "OldSprite"
			await get_tree().process_frame
			var tails = tailsAnimations.instantiate()
			add_child(tails)
			sprite = tails.get_node("Sprite2D")
			animator = tails.get_node("PlayerAnimation")
			superAnimator = tails.get_node_or_null("SuperPalette")
			spriteController = tails
			get_node("OldSprite").queue_free()
```
Every character except sonic needs this unless they really do want to be... just sonic, so yeah this is important.
let's go line by line.
`currentHitbox` is the hitbox your character gets, really in my opinion as long as it doesn't look tooooo jarring you can just use sonic's and it'll be alright. If you're character's freakishly tall or something though their sprites might clip into the ceiling by a bit, but many fangames get away with this just fine, and like Dimension said above, it's really hard to fix.
`get_node("Sonic").name = "OldSprite"` Dont touch this, this gets rid of Sonic's sprites.
`await get_tree().process_frame` Knuckles doesn't have this one, but some characters crash without it and there really isn't a difference whether it's there or not, so just make sure your block uses it and it should be fine.
`var tails = tailsAnimations.instantiate()` THIS is the important thing, switch tailsAnimations into whatever thing you named the variable from the last paragraph to. As for naming the new temporary variable after the character, it's just aestetic, if you want to follow it, you can, you don't have to, but these lines you do have to change to reference the correct one:
```
add_child(tails)
sprite = tails.get_node("Sprite2D")
animator = tails.get_node("PlayerAnimation")
superAnimator = tails.get_node_or_null("SuperPalette")
spriteController = tails
```
As for the last line, it just prepares Sonic's sprites for removal from the scene tree, so don't change nothing.

And... that's it! Your character is still missing their ability that they'd normally have in Air.gd, but that's going to be on you to program honestly, maybe make another thread or reply for that.

Man... I should put this on the wiki...
**FERB, I KNOW WHAT WE'RE GONNA DO TODAY**
